featurebyte/utils/async_helper.py

```diff
@@ -5,18 +5,219 @@
+
+        # Semaphore not acquired - clear identifier from identity and owner zsets
+        pipeline.zrem(self._key, identifier)
+        pipeline.zrem(self._czset, identifier)
+        pipeline.execute()
+        return None
+
+    async def _release_semaphore(self) -> None:
+        """
+        Release semaphore
+        """
+        if not self._identifier:
+            return
+
+        # Release semaphore
+        pipeline = self._redis.pipeline(True)
+        pipeline.zrem(self._key, self._identifier)
+        pipeline.zrem(self._czset, self._identifier)
+        pipeline.execute()
+        self._identifier = None
+
+    async def refresh(self) -> bool:
+        """
+        Refresh semaphore
+
+        Returns
+        -------
+        bool
+            True if semaphore is refreshed, False
+        """
+        if not self._identifier:
+            return False
+
+        # Refresh semaphore
+        if self._redis.zadd(self._key, {self._identifier: time.time()}):
+            # semaphore is lost - clean up
+            await self._release_semaphore()
+            return False
+
+        # Semaphore refreshed
+        return True
+
+    async def __aenter__(self) -> RedisFairCountingSemaphore:
+        """
+        Acquire semaphore on enter, blocks until semaphore is acquired or timeout
+
+        Raises
+        ------
+        TimeoutError
+            If semaphore is not acquired within timeout
+
+        Returns
+        -------
+        RedisFairCountingSemaphore
+        """
+
+        # Try to acquire semaphore
+        now = time.time()
+        while time.time() - now < self._timeout:
+            # Acquire lock for semaphore
+            lock = self._redis.lock(self._lock, timeout=1)
+            if lock.acquire(blocking=False):
+                try:
+                    identifier = await self._acquire_semaphore()
+                    if identifier:
+                        self._identifier = identifier
+                        return self
+                finally:
+                    # Release lock
+                    if lock.owned():
+                        lock.release()
+            await asyncio.sleep(0.1)
+        # Timeout trying to acquire semaphore
+        raise TimeoutError(f"Failed to acquire semaphore after {time.time() - now}s")
+
+    async def __aexit__(self, *exc: dict[str, Any]) -> None:
+        """
+        Release semaphore on exit
+
+        Parameters
+        ----------
+        exc: dict[str, Any]
+            parameters
+        """
+        await self._release_semaphore()
+
+
+async def _semaphore_refresh(task: Task[Any], semaphore: RedisFairCountingSemaphore) -> None:
+    """
+    Refresh semaphore while task is running
+
+    Parameters
+    ----------
+    task: Task[Any]
+        Task to monitor
+    semaphore: RedisFairCountingSemaphore
+        Semaphore to refresh
+    """
+    # Refresh semaphore while task is running
+    while not task.done() and not task.cancelled():
```

Comment:
If the task is an IO task, just want to check whether `task.done()` & `task.cancelled()` works properly in gevent environment.

Reply:
It should. Gevent patches blocking system calls which should not be used by asyncio.
I am also going to refactor this so the refresh task in handled in side the semaphore class, which should no longer require these checks as well